src/transformers/cache_utils.py

```diff
@@ -1339,3 +1339,255 @@ def update_ssm_state(self, layer_idx: int, new_ssm_state: torch.Tensor):
+
+        Return:
+            A tuple containing the updated key and value states.
+        """
+
+        if layer_idx == 0:
+            # Update seen tokens.
+            self._seen_tokens += key_states.shape[-2]
+
+            # Always there.
+            k_out = self.key_cache[0]
+            v_out = self.value_cache[0]
+        else:
+            # Wait for prefetch stream.
+            if self._prefetch_stream is not None:
+                torch.cuda.default_stream(self.device).wait_stream(self._prefetch_stream)
+
+            k_out = self._device_key_cache[layer_idx & 1]
+            v_out = self._device_value_cache[layer_idx & 1]
+
+        self._prefetch_layer(layer_idx + 1)
+
+        cache_position = cache_kwargs.get("cache_position") if cache_kwargs is not None else None
+        if cache_position is None:
+            k_out.copy_(key_states)
+            v_out.copy_(value_states)
+
+            # Copy the values to the offloaded device as well.
+            if layer_idx == 0:
+                self.key_cache[layer_idx].copy_(key_states.to(self.offload_device))
+                self.value_cache[layer_idx].copy_(value_states.to(self.offload_device))
+        else:
+            # Note: here we use `tensor.index_copy_(dim, index, tensor)` that is equivalent to
+            # `tensor[:, :, index] = tensor`, but the first one is compile-friendly and it does
+            # explicitly an in-place operation, that avoids copies and uses less memory.
+            try:
+                k_out.index_copy_(2, cache_position, key_states)
+                v_out.index_copy_(2, cache_position, value_states)
+            except NotImplementedError:
+                # The operator 'aten::index_copy.out' is not currently implemented for the MPS
+                # device.
+                k_out[:, :, cache_position] = key_states
+                v_out[:, :, cache_position] = value_states
+
+            # Copy the values to the offloaded device as well.
+            if layer_idx != 0:
+                cache_position = cache_position.to(self.offload_device)
+                key_states = key_states.to(self.offload_device)
+                value_states = value_states.to(self.offload_device)
+
+                try:
+                    self.key_cache[layer_idx].index_copy_(2, cache_position, key_states)
+                    self.value_cache[layer_idx].index_copy_(2, cache_position, value_states)
+                except NotImplementedError:
+                    # The operator 'aten::index_copy.out' is not currently implemented for the MPS
+                    # device.
+                    self.key_cache[layer_idx][:, :, cache_position] = key_states
+                    self.value_cache[layer_idx][:, :, cache_position] = value_states
+
+        return k_out, v_out
+
+    def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
+        """Returns the sequence length of the cached states that were seen by the model."""
+
+        return self._seen_tokens
```

Comment:
let's use the fn from the static cache (since we want to remove `self._seen_tokens`)

Reply:
That would be a performance degradation compared to the current integer update of `self._seen_tokens`. Is there a plan to remove `get_seq_length` ? In that case, I would remove it then. Otherwise, this will be a lot slower since it will have to do (synced) CPU operation on the offloaded cache. Let me know what you think.

Reply:
Yeah get_seq_length is deprecated in favor of the `cache_positions` which should not need CPU operations as you only use them on device

Reply:
Okay, but shouldn't we just remove the `get_seq_length` methods including all neccesary variables (`_seen_tokens`) once it gets removed from the API? Since there is still quite a few usages throughout the codebase.

What I meant was that `get_seq_length` will be much less performant in the meantime if I switch it over to the `StaticCache` implementation. Which if you want that in the meantime, I'm fine with as well.

Reply:
Ah sorry, no need to use the one from static, and actually yeah, we should probably just prevent user from using it -> no offloading for them? 
Let's go with keeping seentoken for now, add a comment saying it will be deprecated

Reply:
Done. :+1:

Reply:
> we should probably just prevent user from using it -> no offloading for them?
The method `get_seq_length` works fine but it's used internally still. Hence my hesitation to remove it / revert it to the slower `StaticCache` version.

Reply:
Sounds good! 

Me and Arthur will handle the deprecation of both on all cache types + internal usage afterwards 🤗

Reply:
@ArthurZucker Please review and merge when happy.